Use native datetime value in Brother uptime sensor

bieniu opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
As introduced in #52671, sensors now should return a native date(time) when using the date(time) device classes.

This PR:
- adjusts sensor platform to comply with that
- moves consts and classes related to sensor to the sensor platform

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

bieniu committed:
Use datatime as uptime sensor state

bieniu committed:
Use datatime as uptime sensor state

bieniu committed:
Update tests

bieniu committed:
Update tests

micha91 committed:
Add Yamaha MusicCast zone specific devices (#58285)

epenet committed:
Update ZeroconfServiceInfo in tests (a-f) (#60212)

Co-authored-by: epenet <epenet@users.noreply.github.com>

frenck committed:
Adjust CI to split tests into 6 groups (#60198)

emontnemery committed:
Bump pychromecast to 10.1.1 (#60214)

epenet committed:
Update ZeroconfServiceInfo in tests (g-m) (#60217)

Co-authored-by: epenet <epenet@users.noreply.github.com>

epenet committed:
Use BaseServiceInfo inheritance in MqttServiceInfo (#60207)

Co-authored-by: epenet <epenet@users.noreply.github.com>

raman325 committed:
Bump zwave-js-server-python to 0.33.0 (#60213)

mib1185 committed:
Use native datetime value in Shelly sensors (#60179)

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

frenck committed:
Add button support to Alexa (#60163)

epenet committed:
Update ZeroconfServiceInfo in tests (n-t) (#60219)

Co-authored-by: epenet <epenet@users.noreply.github.com>

epenet committed:
Use dataclass for UsbServiceInfo (#60140)

Co-authored-by: epenet <epenet@users.noreply.github.com>

epenet committed:
Update ZeroconfServiceInfo in tests (t-z) (#60221)

Co-authored-by: epenet <epenet@users.noreply.github.com>

bdraco committed:
Ensure homekit setup messages run in event loop (#60226)

bdraco committed:
Fix user input malformed with deleted entities in HomeKit exclude flow (#60061)

bdraco committed:
Split up yeelight code into multiple files (#59990)

cereal2nd committed:
Fully migrate to attribute shorthand in velbus (#59797)

* Move velbus completly to _attr instead of propertys

* Commit all sugestions

* One more sugestion

* Fixed light.py

mib1185 committed:
Use native datetime value in Synology DSM sensors (#60176)

mib1185 committed:
Use native datetime value in AVM Fritz!Tools sensors (#60233)

scop committed:
Remove unneeded pylint disables (#60235)

werkstrom committed:
Add daikin sensor state class (#60145)

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

mib1185 committed:
Use native datetime value in Systemmonitor sensors (#60236)

asymworks committed:
Avoid errors when AirNow API does not return all expected pollutants (#60232)

Co-authored-by: Aaron Bach <bachya1208@gmail.com>

bash-worth committed:
Tado ignore fix (#60011)

Co-authored-by: J. Nick Koston <nick@koston.org>

spacegaier committed:
Consistent capitalization for Synology DSM entity names (#60237)

epenet committed:
Use MqttServiceInfo in tasmota (#60113)

Co-authored-by: epenet <epenet@users.noreply.github.com>

frenck committed:
Use native datetime value in MQTT sensors (#59923)

bdraco committed:
Limit homekit sources to prevent exceeding 100 limit (#59743)

slackr31337 committed:
Bump librouteros version to 3.2.0 (#60066)

sergeymaysak committed:
Bumped version of wirelesstagpy to 0.8.0 (#60125)

emontnemery committed:
Bump device registry version to 1.2 (#60199)

bdraco committed:
Reduce flux_led light turn on complexity (#60139)

cdce8p committed:
Enable strict typing - wallbox (#59301)

avee87 committed:
Add device registry entry for MetOffice (#58683)

emontnemery committed:
Enable basic type checking for awair (#55046)

bdraco committed:
Show how user input is malformed in the UI on error (#60057)

mib1185 committed:
Use native datetime value in IPP sensors (#60234)

epenet committed:
Use dataclass for ZeroconfServiceInfo (#60206)

Co-authored-by: epenet <epenet@users.noreply.github.com>

andreasbrett committed:
Add pi_hole entity "available_updates" (#56181)

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

mib1185 committed:
Use attribute shortcuts in Samsung TV integration (#60171)

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

emontnemery committed:
Use pychromecast CastInfo type in cast integration (#60205)

frenck commented on homeassistant/components/brother/sensor.py:
```diff
@@ -66,13 +130,8 @@ def __init__(
-                getattr(self.coordinator.data, self.entity_description.key).isoformat(),
-            )
         return cast(
             StateType, getattr(self.coordinator.data, self.entity_description.key)
```
```suggestion
            StateType | datetime, getattr(self.coordinator.data, self.entity_description.key)
```

bieniu replied:
Of course, you're right.

bieniu committed:
Add LED brightness for Xiaomi Miio ZA5 fan (#60134)

chishm committed:
dlna_dmr: Ignore philips tv (#60204)

bieniu committed:
Use native datetime value in NAM uptime sensor (#60241)

cvroque committed:
Check if Tuya Vacuum (sd) is able to report status directly before using pause switch (#59888)

kamildoleglo committed:
Fix Tuya integration for climate devices (#60229)

skynetua committed:
Add hjjcy device category to Tuya integration (#60224)

frenck committed:
Run partial test suite in CI if core untouched (#60230)

Co-authored-by: J. Nick Koston <nick@koston.org>
Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

frenck committed:
Make partial test suite not waiting for linters (#60254)

chemelli74 committed:
Add slow tests list to VScode task code coverage (#60183)

frenck committed:
Add partial codecov flag to CI to trigger carryforward coverage (#60256)

farmio committed:
Rewrite Fronius integration (#59686)

* Add unique_id and use DataUpdateCoordinator in Fronius (#57879)

* initial refactoring commit - meters

- config_flow (no strings, no tests yet)
- import yaml config
- FroniusSolarNet class for holding Fronius object , coordinators and some common data
- meter descriptions
- update coordinator
- entities (including devices)

* storage controllers

* error handling on init; inverter unique_id

* inverters

* power_flow

* fix VA, var, varh not valid for device_class power/energy

and add custom icons

* add SolarNet device for system wide values

* cleanup

* config_flow strings

* test config_flow

* use pyfronius 0.7.0

* enable strict typing

* remove TODO comments

* fix lint errors; move FroniusSensorEntity to sensor.py

* power_flow as optional coordinator

API V0 doesn't support power_flow endpoint

* show error message in logs

* prevent parallel requests to one host

* logger_info coordinator

* store FroniusSolarNet reference directly in coordinator

* cleanup coordinators when unloading entry

* round floats returned by Fronius API

* default icons for grid im/export tariffs

* small typing fix

* Update homeassistant/components/fronius/sensor.py

Co-authored-by: Brett Adams <Bre77@users.noreply.github.com>

* DC icons

* prepend names with "Fronius" and device type

to get more reasonable default entity_ids (eg. have them next to each other when alphabetically sorted)

* remove config_flow and devices

* rename _FroniusUpdateCoordinator to FroniusCoordinatorBase

and mark ABC

* move SensorEntityDescriptions to sensor.py

* Revert "move SensorEntityDescriptions to sensor.py"

This reverts commit 2e5a726eb65854f236a0c72f3f67f04a6f8a2eff.

* Don't raise ConfigEntryNotReady and use regular refresh method

* move bridge initialization out of helper class

* no coverage tests

* power_flow update interval 10 seconds

* move SensorEntityDescriptions to sensor.py

without introducing a circular dependency

* deprecation warning for CONF_MONITORED_CONDITIONS

* remove extra_state_attributes form meter sensor entities

* readd diagnostic entities

* decouple default entity_id from default name

* use key instead of name for entity_id

and make deprecated config key optional

* adjust tests

* use old entity_ids

these changes are now backwards compatible

* check coverage

* simplify entity description definitions

* restore entity names of previous implementation

Co-authored-by: Brett Adams <Bre77@users.noreply.github.com>

* Add config_flow for Fronius integration (#59677)

* Cleanup Fronius config_flow and tests (#60094)

* Add devices to Fronius integration (#60104)

* New entity names for Fronius entities (#60215)

* Adaptive update interval for Fronius coordinators (#60192)

Co-authored-by: Brett Adams <Bre77@users.noreply.github.com>

frenck committed:
Temporary disable partial runs in CI (#60258)

cvroque committed:
Add configuration entities to Tuya Vacuum (sd) (#59936)

chemelli74 committed:
Fix socket usage in Aprs test (#60253)

bdraco committed:
Prevent get_mac_address from blocking event loop in samsungtv config flow (#60246)

felipediel committed:
Decouple BroadlinkDevice.available property (#58853)

* Decouple BroadlinkDevice.available property

* Exclude unreachable line from coverage

emontnemery committed:
Add stable ID to entity registry entries (#60218)

* Add UUID to entity registry entries

* Fix test

epenet committed:
Use UsbServiceInfo in modem_callerid (#60268)

Co-authored-by: epenet <epenet@users.noreply.github.com>

pvizeli committed:
Update base image 2021.11.0 (#60227)

jbouwh committed:
Change output template filters `timestamp_local` and `timestamp_utc` to isoformat (#60269)

jbouwh committed:
Add UNIX timestamp detection to `as_datetime` template filter (#60126)

epenet committed:
Revert "Add type hints to SSDP (#59840)" (#60270)

frenck committed:
Correct entity category on UPS type in NUT (#60277)

emontnemery committed:
Add comments to entity category constants (#60276)

farmio committed:
Disable less popular Fronius entities by default (#60264)

chemelli74 committed:
Fix init slow tests for SamsungTV (#60245)

mkowalchuk committed:
Use configured speed ranges for HomeSeer FC200+ fan controllers in zwave_js (#59697)

* Use configured speed ranges for HomeSeer FC200+ fan controllers in zwave_js

* Fix pylint errors

* Remove unused param in tests

* Fix test values

* Address various review notes

* Remove now-redundant assertion

* Add an additional test case for set_percentage=0

* Use round() instead of int() for percentage computations; this makes the percentage setting match the setpoints in the UI

* Add additional tests

* Fix pct conversions

* Make conversion tests exhaustive

* Add tests for discovery data templates

* Revert "Add tests for discovery data templates"

This reverts commit 85dcbc0903a1dd95f8e4e5f3c5d29cd7547b667b.

* Improve typing on ConfigurableFanSpeedDataTemplate#resolve_data

* Move config error handling to the discovery data template

* Fix checks for config data

* Revise fallback logic in percentage_to_zwave_speed and ensure that the speed list is non-empty

* Rework error handling

* Fix runtime fan speed updates

* Use warning instead of warn

* Move data validation to get_speed_config; turns out that resolve_data is only called once, at startup.

* Temporarily remove the not-yet-used fixed fan speed template.  Add an additional assertion to ensure speeds are sorted.

* Add a comment about the assertions in discovery_data_template.py

* Update homeassistant/components/zwave_js/discovery_data_template.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

* Fix typo in comment

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

h2zero committed:
Fix Konnected multiple discovery of panels (#59953)

* Konnected - Fix multiple discovery of panels.

This resolves an issue which creates multiple discoveries of a Konnected panel if it is restarted and fails to connect to home assistant.

See #57467.

* Revert changes to user step, add handling to ssdp step.

* Add abort reason string to strings.json

* Abort ssdp discovery if device is already known.

* Add test for multiple discovery fix.

* Remove unrelated file change.

* Add ssdp discovery abort tests.

* Add missing abort reason check.

* Add "already_configured" to strings.

* Use "cannot_connect" abort reason.

frenck committed:
CI: Combine smaller linters into a single job (#60279)

* CI: Combine smaller linters into a single job

* Remove unrelated changhe

* Remove stale exit

ThomDietrich committed:
Add additional statistics characteristics, remove attributes (#59867)

* Add additional statistics characterics, improve rounding

* Improve name of age_usage_ratio

* Replace difference by three relevant distances

* Refactor attributes, remove stats, add metadata

* Fix binary sensor testcase

* Fix sensor defaults testcase

* Fix and enhance all testcases

* Remove age coverage from attr when not configured

* Refactor so only the relevant characteristic value is calculated

* Rename unclear characteristics, add timebound average

* Fix coverage warning

* Remove explicit functions dict

epenet committed:
Update ssdp matching_domain constant (#60283)

Co-authored-by: epenet <epenet@users.noreply.github.com>

epenet committed:
Use UsbServiceInfo and ZeroconfServiceInfo in zha (#60266)

Co-authored-by: epenet <epenet@users.noreply.github.com>

epenet committed:
Use UsbServiceInfo in zwave-js (#60267)

Co-authored-by: epenet <epenet@users.noreply.github.com>

frenck committed:
Use UTC timestamp in uptime sensor (#60240)

jbouwh committed:
Add bytes support for bitwise template operations (#60211)

* Add bytes support for bitwise template operations

* spelling

* Update bitwise tests

* remove try block for bytes conversion

* do not accept empty `bytes` object

emontnemery committed:
Correct today_at template function / filter (#60291)

Co-authored-by: Franck Nijhof <git@frenck.dev>

frenck committed:
Partial CI workflows: take 2 (#60294)

frenck committed:
CI: Fix full suite on core file changes (#60299)

frenck committed:
CI: Fix test groups when running full CI suite (#60300)

frenck committed:
Use native datetime value in OpenWeatherMap sensors (#60303)

bdraco committed:
Add support for adjusting effect brightness in flux_led (#60247)

bdraco committed:
Refactor zeroconf matching to be more DRY (#60293)

emontnemery committed:
Allow MQTT selects to have a single or no options (#60281)

FaserF committed:
Bump openwrt-luci-rpc from 1.1.8 to 1.1.11 (#60124)

bieniu committed:
Update homeassistant/components/brother/sensor.py

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

bieniu committed:
Merge branch 'brother/uptime' of https://github.com/bieniu/home-assistant-core into brother/uptime

bieniu committed:
Fix typing

bieniu commented:
Ooops, sorry for notifications :D
